File: app/src/test/java/piuk/blockchain/android/ui/swipetoreceive/SwipeToReceiveHelperTest.java
package piuk.blockchain.android.ui.swipetoreceive;

import info.blockchain.api.data.Balance;
import info.blockchain.wallet.payload.data.Account;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.math.BigInteger;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.observers.TestObserver;
import piuk.blockchain.android.RxTest;
import piuk.blockchain.android.data.datamanagers.PayloadDataManager;
import piuk.blockchain.android.util.PrefsUtil;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.KEY_SWIPE_RECEIVE_ACCOUNT_NAME;
import static piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.KEY_SWIPE_RECEIVE_ADDRESSES;

public class SwipeToReceiveHelperTest extends RxTest {

    private SwipeToReceiveHelper subject;
    @Mock private PayloadDataManager payloadDataManager;
    @Mock private PrefsUtil prefsUtil;

    @Before
    public void setUp() throws Exception {
        super.setUp();
        MockitoAnnotations.initMocks(this);

        subject = new SwipeToReceiveHelper(payloadDataManager, prefsUtil);
    }

    @Test
    public void updateAndStoreAddresses() throws Exception {
        // Arrange
        when(prefsUtil.getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true)).thenReturn(true);
        Account mockAccount = mock(Account.class);
        when(payloadDataManager.getDefaultAccount()).thenReturn(mockAccount);
        when(mockAccount.getLabel()).thenReturn("Account");
        when(payloadDataManager.getReceiveAddressAtPosition(eq(mockAccount), anyInt())).thenReturn("address");
        // Act
        subject.updateAndStoreAddresses();
        // Assert
        verify(prefsUtil).getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true);
        verify(payloadDataManager, times(5)).getReceiveAddressAtPosition(eq(mockAccount), anyInt());
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_ACCOUNT_NAME, "Account");
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_ADDRESSES, "address,address,address,address,address,");
    }

    @Test
    public void getNextAvailableAddressValid() throws Exception {
        // Arrange
        LinkedHashMap<String, Balance> map = new LinkedHashMap<>();
        Balance balance0 = new Balance();
        Balance balance1 = new Balance();
        Balance balance2 = new Balance();
        Balance balance3 = new Balance();
        Balance balance4 = new Balance();
        balance0.setFinalBalance(BigInteger.valueOf(1000L));
        balance1.setFinalBalance(BigInteger.valueOf(5L));
        balance2.setFinalBalance(BigInteger.valueOf(-10L));
        balance3.setFinalBalance(BigInteger.valueOf(0L));
        balance4.setFinalBalance(BigInteger.valueOf(0L));
        map.put("addr0", balance0);
        map.put("addr1", balance1);
        map.put("addr2", balance2);
        map.put("addr3", balance3);
        map.put("addr4", balance4);
        when(payloadDataManager.getBalanceOfAddresses(anyList())).thenReturn(Observable.just(map));
        when(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, "")).thenReturn("addr0, addr1, addr2, addr3, addr4");
        // Act
        TestObserver<String> testObserver = subject.getNextAvailableAddress().test();
        // Assert
        testObserver.assertComplete();
        testObserver.assertNoErrors();
        testObserver.assertValue("addr3");
    }

    @Test
    public void getNextAvailableAddressAllUsed() throws Exception {
        // Arrange
        LinkedHashMap<String, Balance> map = new LinkedHashMap<>();
        Balance balance0 = new Balance();
        Balance balance1 = new Balance();
        Balance balance2 = new Balance();
        Balance balance3 = new Balance();
        Balance balance4 = new Balance();
        balance0.setFinalBalance(BigInteger.valueOf(1000L));
        balance1.setFinalBalance(BigInteger.valueOf(5L));
        balance2.setFinalBalance(BigInteger.valueOf(-10L));
        balance3.setFinalBalance(BigInteger.valueOf(1L));
        balance4.setFinalBalance(BigInteger.valueOf(1_000_000_000_000L));
        map.put("addr0", balance0);
        map.put("addr1", balance1);
        map.put("addr2", balance2);
        map.put("addr3", balance3);
        map.put("addr4", balance4);
        when(payloadDataManager.getBalanceOfAddresses(anyList())).thenReturn(Observable.just(map));
        when(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, "")).thenReturn("addr0, addr1, addr2, addr3, addr4");
        // Act
        TestObserver<String> testObserver = subject.getNextAvailableAddress().test();
        // Assert
        testObserver.assertComplete();
        testObserver.assertNoErrors();
        testObserver.assertValue("");
    }

    @Test
    public void getReceiveAddresses() throws Exception {
        // Arrange
        when(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, "")).thenReturn("addr0, addr1, addr2, addr3, addr4");
        // Act
        List<String> result = subject.getReceiveAddresses();
        // Assert
        assertEquals(5, result.size());
    }

    @Test
    public void getReceiveAddressesEmptyList() throws Exception {
        // Arrange
        when(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, "")).thenReturn("");
        // Act
        List<String> result = subject.getReceiveAddresses();
        // Assert
        assertEquals(Collections.emptyList(), result);
    }

    @Test
    public void getAccountName() throws Exception {
        // Arrange
        when(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ACCOUNT_NAME, "")).thenReturn("Account");
        // Act
        String result = subject.getAccountName();
        // Assert
        assertEquals("Account", result);
    }

}

File: app/src/test/java/piuk/blockchain/android/ui/swipetoreceive/SwipeToReceiveHelperTest.kt
package piuk.blockchain.android.ui.swipetoreceive

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import info.blockchain.api.data.Balance
import info.blockchain.wallet.coin.GenericMetadataAccount
import info.blockchain.wallet.payload.data.Account
import io.reactivex.Observable
import org.amshove.kluent.`should equal`
import org.bitcoinj.params.BitcoinCashMainNetParams
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyList
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito
import org.mockito.Mockito.atLeastOnce
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import piuk.blockchain.android.R
import piuk.blockchain.android.RxTest
import piuk.blockchain.android.data.bitcoincash.BchDataManager
import piuk.blockchain.android.data.ethereum.EthDataManager
import piuk.blockchain.android.data.payload.PayloadDataManager
import piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.Companion.KEY_SWIPE_RECEIVE_ACCOUNT_NAME
import piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.Companion.KEY_SWIPE_RECEIVE_ADDRESSES
import piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.Companion.KEY_SWIPE_RECEIVE_BCH_ACCOUNT_NAME
import piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.Companion.KEY_SWIPE_RECEIVE_BCH_ADDRESSES
import piuk.blockchain.android.ui.swipetoreceive.SwipeToReceiveHelper.Companion.KEY_SWIPE_RECEIVE_ETH_ADDRESS
import piuk.blockchain.android.util.NetworkParameterUtils
import piuk.blockchain.android.util.PrefsUtil
import piuk.blockchain.android.util.StringUtils
import java.math.BigInteger
import java.util.*

class SwipeToReceiveHelperTest : RxTest() {

    private lateinit var subject: SwipeToReceiveHelper
    private val payloadDataManager: PayloadDataManager = mock()
    private val prefsUtil: PrefsUtil = mock()
    private val stringUtils: StringUtils = mock()
    private val ethDataManager: EthDataManager = mock(defaultAnswer = Mockito.RETURNS_DEEP_STUBS)
    private val bchDataManager: BchDataManager = mock()
    private val networkParameterUtils: NetworkParameterUtils = mock()

    @Before
    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()

        whenever(networkParameterUtils.bitcoinCashParams).thenReturn(BitcoinCashMainNetParams.get())

        subject = SwipeToReceiveHelper(
                payloadDataManager,
                prefsUtil,
                ethDataManager,
                bchDataManager,
                stringUtils,
                networkParameterUtils
        )
    }

    @Test
    @Throws(Exception::class)
    fun updateAndStoreBitcoinAddresses() {
        // Arrange
        whenever(prefsUtil.getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true))
                .thenReturn(true)
        val mockAccount: Account = mock()
        whenever(payloadDataManager.defaultAccount).thenReturn(mockAccount)
        whenever(mockAccount.label).thenReturn("Account")
        whenever(payloadDataManager.getReceiveAddressAtPosition(eq(mockAccount), anyInt()))
                .thenReturn("address")
        // Act
        subject.updateAndStoreBitcoinAddresses()
        // Assert
        verify(prefsUtil).getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true)
        verify(payloadDataManager, times(5)).getReceiveAddressAtPosition(eq(mockAccount), anyInt())
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_ACCOUNT_NAME, "Account")
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_ADDRESSES, "address,address,address,address,address,")
    }

    @Ignore("Currently our implementation of BECH32 for BCH fails")
    @Test
    @Throws(Exception::class)
    fun updateAndStoreBitcoinCashAddresses() {
        // Arrange
        whenever(prefsUtil.getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true))
                .thenReturn(true)
        val mockAccount: GenericMetadataAccount = mock()
        whenever(bchDataManager.getDefaultGenericMetadataAccount()).thenReturn(mockAccount)
        whenever(bchDataManager.getDefaultAccountPosition()).thenReturn(0)
        whenever(mockAccount.label).thenReturn("BCH Account")
        whenever(bchDataManager.getReceiveAddressAtPosition(eq(0), anyInt()))
                .thenReturn("bitcoincash:qpm2qsznhks23z7629mms6s4cwef74vcwvy22gdx6a")
        // Act
        subject.updateAndStoreBitcoinCashAddresses()
        // Assert
        verify(prefsUtil).getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true)
        verify(bchDataManager, times(5)).getReceiveAddressAtPosition(eq(0), anyInt())
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_BCH_ACCOUNT_NAME, "BCH Account")
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_BCH_ADDRESSES, "1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,")
    }

    @Test
    @Throws(Exception::class)
    fun storeEthAddress() {
        // Arrange
        whenever(prefsUtil.getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true))
                .thenReturn(true)
        whenever(ethDataManager.getEthWallet()?.account?.address).thenReturn("address")
        // Act
        subject.storeEthAddress()
        // Assert
        verify(prefsUtil).getValue(PrefsUtil.KEY_SWIPE_TO_RECEIVE_ENABLED, true)
        verify(ethDataManager, atLeastOnce()).getEthWallet()
        verify(prefsUtil).setValue(KEY_SWIPE_RECEIVE_ETH_ADDRESS, "address")
    }

    @Test
    @Throws(Exception::class)
    fun getNextAvailableAddressSingleValid() {
        // Arrange
        val map = LinkedHashMap<String, Balance>()
        val balance0 = Balance().apply { finalBalance = BigInteger.valueOf(1000L) }
        val balance1 = Balance().apply { finalBalance = BigInteger.valueOf(5L) }
        val balance2 = Balance().apply { finalBalance = BigInteger.valueOf(-10L) }
        val balance3 = Balance().apply { finalBalance = BigInteger.valueOf(0L) }
        val balance4 = Balance().apply { finalBalance = BigInteger.valueOf(0L) }
        map.put("addr0", balance0)
        map.put("addr1", balance1)
        map.put("addr2", balance2)
        map.put("addr3", balance3)
        map.put("addr4", balance4)
        whenever(payloadDataManager.getBalanceOfAddresses(anyList()))
                .thenReturn(Observable.just(map))
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, ""))
                .thenReturn("addr0, addr1, addr2, addr3, addr4")
        // Act
        val testObserver = subject.getNextAvailableBitcoinAddressSingle().test()
        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertValue("addr3")
    }

    @Test
    @Throws(Exception::class)
    fun getNextAvailableAddressSingleAllUsed() {
        // Arrange
        val map = LinkedHashMap<String, Balance>()
        val balance0 = Balance().apply { finalBalance = BigInteger.valueOf(1000L) }
        val balance1 = Balance().apply { finalBalance = BigInteger.valueOf(5L) }
        val balance2 = Balance().apply { finalBalance = BigInteger.valueOf(-10L) }
        val balance3 = Balance().apply { finalBalance = BigInteger.valueOf(1L) }
        val balance4 = Balance().apply { finalBalance = BigInteger.valueOf(1_000_000_000_000L) }
        map.put("addr0", balance0)
        map.put("addr1", balance1)
        map.put("addr2", balance2)
        map.put("addr3", balance3)
        map.put("addr4", balance4)
        whenever(payloadDataManager.getBalanceOfAddresses(anyList()))
                .thenReturn(Observable.just(map))
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, ""))
                .thenReturn("addr0, addr1, addr2, addr3, addr4")
        // Act
        val testObserver = subject.getNextAvailableBitcoinAddressSingle().test()
        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertValue("")
    }

    @Ignore("Currently our implementation of BECH32 for BCH fails")
    @Test
    @Throws(Exception::class)
    fun `getNextAvailableBCHAddressSingle Valid`() {
        // Arrange
        val map = LinkedHashMap<String, Balance>()
        val balance0 = Balance().apply { finalBalance = BigInteger.valueOf(1000L) }
        val balance1 = Balance().apply { finalBalance = BigInteger.valueOf(5L) }
        val balance2 = Balance().apply { finalBalance = BigInteger.valueOf(-10L) }
        val balance3 = Balance().apply { finalBalance = BigInteger.valueOf(0L) }
        val balance4 = Balance().apply { finalBalance = BigInteger.valueOf(0L) }
        map.put("1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu", balance0)
        map.put("1KXrWXciRDZUpQwQmuM1DbwsKDLYAYsVLR", balance1)
        map.put("16w1D5WRVKJuZUsSRzdLp9w3YGcgoxDXb", balance2)
        map.put("3CWFddi6m4ndiGyKqzYvsFYagqDLPVMTzC", balance3)
        map.put("3LDsS579y7sruadqu11beEJoTjdFiFCdX4", balance4)
        whenever(payloadDataManager.getBalanceOfBchAddresses(anyList()))
                .thenReturn(Observable.just(map))
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_BCH_ADDRESSES, ""))
                .thenReturn("1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,1KXrWXciRDZUpQwQmuM1DbwsKDLYAYsVLR,16w1D5WRVKJuZUsSRzdLp9w3YGcgoxDXb,3CWFddi6m4ndiGyKqzYvsFYagqDLPVMTzC,3LDsS579y7sruadqu11beEJoTjdFiFCdX4")
        // Act
        val testObserver = subject.getNextAvailableBitcoinCashAddressSingle().test()
        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertValue("bitcoincash:ppm2qsznhks23z7629mms6s4cwef74vcwvn0h829pq")
    }

    @Ignore("Currently our implementation of BECH32 for BCH fails")
    @Test
    @Throws(Exception::class)
    fun `getNextAvailableBCHAddressSingle All Used`() {
        // Arrange
        val map = LinkedHashMap<String, Balance>()
        val balance0 = Balance().apply { finalBalance = BigInteger.valueOf(1000L) }
        val balance1 = Balance().apply { finalBalance = BigInteger.valueOf(5L) }
        val balance2 = Balance().apply { finalBalance = BigInteger.valueOf(-10L) }
        val balance3 = Balance().apply { finalBalance = BigInteger.valueOf(1L) }
        val balance4 = Balance().apply { finalBalance = BigInteger.valueOf(1_000_000_000_000L) }
        map.put("1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu", balance0)
        map.put("1KXrWXciRDZUpQwQmuM1DbwsKDLYAYsVLR", balance1)
        map.put("16w1D5WRVKJuZUsSRzdLp9w3YGcgoxDXb", balance2)
        map.put("3CWFddi6m4ndiGyKqzYvsFYagqDLPVMTzC", balance3)
        map.put("3LDsS579y7sruadqu11beEJoTjdFiFCdX4", balance4)
        whenever(payloadDataManager.getBalanceOfBchAddresses(anyList()))
                .thenReturn(Observable.just(map))
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_BCH_ADDRESSES, ""))
                .thenReturn("1BpEi6DfDAUFd7GtittLSdBeYJvcoaVggu,1KXrWXciRDZUpQwQmuM1DbwsKDLYAYsVLR,16w1D5WRVKJuZUsSRzdLp9w3YGcgoxDXb,3CWFddi6m4ndiGyKqzYvsFYagqDLPVMTzC,3LDsS579y7sruadqu11beEJoTjdFiFCdX4")
        // Act
        val testObserver = subject.getNextAvailableBitcoinCashAddressSingle().test()
        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertValue("")
    }

    @Test
    @Throws(Exception::class)
    fun getEthReceiveAddressSingle() {
        // Arrange
        val address = "ADDRESS"
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ETH_ADDRESS, null))
                .thenReturn(address)
        // Act
        val testObserver = subject.getEthReceiveAddressSingle().test()
        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertValue(address)
    }

    @Test
    @Throws(Exception::class)
    fun getBitcoinReceiveAddresses() {
        // Arrange
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, ""))
                .thenReturn("addr0, addr1, addr2, addr3, addr4")
        // Act
        val result = subject.getBitcoinReceiveAddresses()
        // Assert
        assertEquals(5, result.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun getBitcoinCashReceiveAddressesEmptyList() {
        // Arrange
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ADDRESSES, ""))
                .thenReturn("")
        // Act
        val result = subject.getBitcoinReceiveAddresses()
        // Assert
        assertEquals(emptyList<Any>(), result)
    }

    @Test
    @Throws(Exception::class)
    fun getBitcoinCashReceiveAddresses() {
        // Arrange
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_BCH_ADDRESSES, ""))
                .thenReturn("addr0, addr1, addr2, addr3, addr4")
        // Act
        val result = subject.getBitcoinCashReceiveAddresses()
        // Assert
        assertEquals(5, result.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun getBitcoinReceiveAddressesEmptyList() {
        // Arrange
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_BCH_ADDRESSES, ""))
                .thenReturn("")
        // Act
        val result = subject.getBitcoinCashReceiveAddresses()
        // Assert
        assertEquals(emptyList<Any>(), result)
    }

    @Test
    @Throws(Exception::class)
    fun getEthReceiveAddress() {
        // Arrange
        val address = "ADDRESS"
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ETH_ADDRESS, null))
                .thenReturn(address)
        // Act
        val result = subject.getEthReceiveAddress()
        // Assert
        verify(prefsUtil).getValue(KEY_SWIPE_RECEIVE_ETH_ADDRESS, null)
        result `should equal` address
    }

    @Test
    @Throws(Exception::class)
    fun getBitcoinAccountName() {
        // Arrange
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_ACCOUNT_NAME, ""))
                .thenReturn("Account")
        // Act
        val result = subject.getBitcoinAccountName()
        // Assert
        result `should equal` "Account"
    }

    @Test
    @Throws(Exception::class)
    fun getBitcoinCashAccountName() {
        // Arrange
        val defaultAccountName = "Default account name"
        whenever(prefsUtil.getValue(KEY_SWIPE_RECEIVE_BCH_ACCOUNT_NAME, defaultAccountName))
                .thenReturn("Account")
        whenever(stringUtils.getString(R.string.bch_default_account_label))
                .thenReturn(defaultAccountName)
        // Act
        val result = subject.getBitcoinCashAccountName()
        // Assert
        result `should equal` "Account"
    }

    @Test
    @Throws(Exception::class)
    fun getEthAccountName() {
        // Arrange
        val label = "LABEL"
        whenever(stringUtils.getString(R.string.eth_default_account_label)).thenReturn(label)
        // Act
        val result = subject.getEthAccountName()
        // Assert
        result `should equal` label
    }

}